neurom/features/morphology.py

```diff
@@ -166,37 +163,206 @@ def trunk_vectors(morph, neurite_type=NeuriteType.all):
 
-    return [morphmath.angle_between_vectors(ordered_vectors[i], ordered_vectors[i - 1])
-            for i, _ in enumerate(ordered_vectors)]
+    # Ensure azimuth differences are in [-2pi, 2pi]
+    angles[:, :, 2] = morphmath.angles_to_pi_interval(angles[:, :, 2], scale=2.0)
+
+    if closest_component is not None:
+        angles = angles[
+            np.arange(len(angles)),
+            np.argmin(np.abs(angles[:, :, closest_component]), axis=1)
+        ][:, np.newaxis, :]
+
+    return angles.tolist()
 
 
 @feature(shape=(...,))
-def trunk_origin_radii(morph, neurite_type=NeuriteType.all):
-    """Radii of the trunk sections of neurites in a morph."""
-    return [n.root_node.points[0][COLS.R]
+def trunk_angles_from_vector(
+    morph,
+    neurite_type=NeuriteType.all,
+    vector=None,
+):
+    """Calculate the angles between the trunks of the morph of a given type and a given vector.
+
+    For each neurite, an array with 3 elements is returned:
+    * the absolute 3d angle between the two vectors.
+    * the elevation angle (or polar angle) between the two vectors.
+    * the azimuth angle between the two vectors.
+
+    If ``vector`` is ``None``, the reference vector is set to `(0, 1, 0)`.
+    """
+    if vector is None:
+        vector = (0, 1, 0)
+
+    vectors = np.array(trunk_vectors(morph, neurite_type=neurite_type))
+
+    # In order to avoid the failure of the process in case the neurite_type does not exist
+    if len(vectors) == 0:
+        return []
+
+    angles = np.empty((len(vectors), 3), dtype=float)
+    for i, i_vec in enumerate(vectors):
+        angles[i, 0] = morphmath.angle_between_vectors(vector, i_vec)
+        angles[i, (1, 2)] = morphmath.spherical_from_vector(i_vec) - morphmath.spherical_from_vector(vector)
+
+    # Ensure elevation difference are in [-pi, pi]
+    angles[:, 1] = morphmath.angles_to_pi_interval(angles[:, 1])
+
+    # Ensure azimuth difference are in [-2pi, 2pi]
+    angles[:, 2] = morphmath.angles_to_pi_interval(angles[:, 2], scale=2)
+
+    return angles.tolist()
+
+
+@feature(shape=(...,))
+def trunk_origin_radii(
+    morph,
+    neurite_type=NeuriteType.all,
+    min_length_filter=None,
+    max_length_filter=None,
+):
+    """Radii of the trunk sections of neurites in a morph.
+
+    If ``min_length_filter`` and / or ``max_length_filter`` is given, the points are filtered and
+    the mean radii of the remaining points is returned.
+    """
+    if max_length_filter is None and min_length_filter is None:
+        return [n.root_node.points[0][COLS.R]
+                for n in iter_neurites(morph, filt=is_type(neurite_type))]
+
+    def _mean_radius(points):
+        interval_lengths = morphmath.interval_lengths(points)
+        path_lengths = np.insert(np.cumsum(interval_lengths), 0, 0)
+        valid_pts = np.ones(len(path_lengths), dtype=bool)
+        if min_length_filter is not None:
+            valid_pts = (valid_pts & (path_lengths >= min_length_filter))
```

Comment:
Is the equality here crucial to check it with `np.isclose`?

Reply:
I don't think `np.close` is needed here.

Reply:
ok, I was wondering.